DRILL-6791: Scan projection framework

paul-rogers opened the pull request:
The "schema projection" mechanism:

* Handles none (SELECT COUNT\(*)), some (SELECT a, b, x) and all (SELECT *) projection.
* Handles null columns (for projection a column "x" that does not exist in the base table.)
* Handles constant columns as used for file metadata (AKA "implicit" columns).
* Handle schema persistence: the need to reuse the same vectors across different scanners
* Provides a framework for consuming externally-supplied metadata
* Since we don't yet have a way to provide "real" metadata, obtains metadata hints from previous batches and from the projection list (a.b implies that "a" is a map, c[0] implies that "c" is an array, etc.)
* Handles merging the set of data source columns and null columns to create the final output batch.

vdiravka commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/ScanTestUtils.java:
```diff
@@ -0,0 +1,77 @@
+   */
+
+  public static List<ScanProjectionParser> parsers(ScanProjectionParser... parsers) {
+    return Lists.newArrayList(parsers);
```
Use `new ArrayList<>()` to make Travis build happy :)  
The note in Guava `ListsnewArrayList()` javadoc:
> <p><b>Note for Java 7 and later:</b> this method is now unnecessary and
>  should be treated as deprecated. Instead, use the {@code ArrayList}
> {@linkplain ArrayList#ArrayList() constructor} directly, taking advantage
> of the new <a href="http://goo.gl/iz2Wi">"diamond" syntax</a>.

paul-rogers replied:
Fixed

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/project/NoOpMetadataManager.java:
```diff
@@ -0,0 +1,54 @@
+
+  @Override
+  public SchemaProjectionResolver resolver() {
+    throw new UnsupportedOperationException();
```
Fixed

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/project/ScanLevelProjection.java:
```diff
@@ -0,0 +1,349 @@
+      // By putting them at the end, the index of all other columns remains
+      // constant. Drill 1.12 broke that behavior, but Drill 1.13 restored it.
+      //
+      // This option can be removed in Drill 1.14 after things settle down.
```
Let's leave it for now. We can test it once we get the revised readers integrated and the test case that triggered this hack will tell us if Drill still behaves as in 1.12, or if the 1.11 and earlier behavior has been restored.

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/ScanTestUtils.java:
```diff
@@ -0,0 +1,77 @@
+
+  // Default file metadata column names; primarily for testing.
+
+  public static final String FILE_NAME_COL = "filename";
```
Tried, but it seems that the class no longer exists. As I recall, the names came from config settings or session options.

Actually, we don't even need these fields for this PR, so I removed them. Will add them in a later PR that adds the file metadata mechanism (the successor to the implicit column explorer.)

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/project/TestScanBatchWriters.java:
```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+package org.apache.drill.exec.physical.impl.scan.project;
```
Fixed

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/project/TestSchemaSmoothing.java:
```diff
@@ -0,0 +1,681 @@
+       assertEquals(4, smoother.schemaVersion());
+    }
+
+//    // Table 6: (a: not-nullable bigint): convert to nullable for consistency
```
Please clean up.

arina-ielchiieva commented on exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/project/TestSchemaSmoothing.java:
```diff
@@ -0,0 +1,681 @@
+    projector.close();
+  }
+
+  // TODO: Test schema smoothing with repeated
```
Not really in favor of submitting code with todos...

arina-ielchiieva commented:
@paul-rogers would you have a chance to address code review comments before release cut off date?

paul-rogers commented:
@arina-ielchiieva, addressed your code review comments. Also, found and fixed a "bits" vector initialization bug. Please take another look to see if there is anything else we need to address.

paul-rogers commented:
Squashed commits and rebased on latest master.

vdiravka commented:
Most likely the Travis build fails as usual due to timeout limit 50mins. I have created PR to fix that.

`TestPStoreProviders.verifyZkStore()` test is marked as `SlowTest`, so it is excluded from Travis build.
I have checked it on your branch and it passed for me, but with error message:
> mvn -Dtest=TestPStoreProviders#verifyZkStore -pl exec/java-exec test
> ...
> [Time-limited test-EventThread] ERROR org.apache.curator.ConnectionState - Authentication failed

The same message is on current master. So it is not related to your changes (but a good point to discuss why it appears).
